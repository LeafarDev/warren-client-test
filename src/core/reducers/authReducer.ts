import * as actions from '../actionTypes/authActionsTypes';

export interface AuthState {
  token: string;
}

const initialState: AuthState = {
  token: ''
};

export default function authReducer(
  state: AuthState = initialState,
  action: actions.AuthAction
): AuthState {
  switch (action.type) {
    case actions.RESET:
    case actions.LOGOUT:
      return initialState;
    case actions.LOGIN_SUCCESS:
      return {
        token: action.token
      };
    default:
      return state;
  }
}
