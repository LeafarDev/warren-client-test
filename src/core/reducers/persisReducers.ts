import storage from 'redux-persist/lib/storage';
import { persistReducer } from 'redux-persist';
import { Reducer } from 'redux';

export default (reducers: Reducer) => {
  const persistedReducer = persistReducer(
    {
      key: 'bankofmars',
      storage,
      whitelist: ['auth']
    },
    reducers
  );

  return persistedReducer;
};
