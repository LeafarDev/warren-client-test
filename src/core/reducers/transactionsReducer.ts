import * as actions from '../actionTypes/transactionsActionTypes';

export interface ITransaction {
  id: string | null
  sourceAccountNumber: string
  targetAccountNumber: string | null
  amount: number
  type: 'withdraw' | 'deposit' | 'payment'
  createdAt: string | null
  updatedAt: string | null
}

export interface TransactionsState {
  transactions: ITransaction[];
  lastCreatedTransaction: ITransaction | null;
  currentCreatingTransactionType: 'withdraw' | 'deposit' | 'payment';
}

const initialState: TransactionsState = {
  transactions: [],
  currentCreatingTransactionType: 'deposit',
  lastCreatedTransaction: null
};

export default function transactionsReducer(
  state: TransactionsState = initialState,
  action: actions.TransactionsAction
): TransactionsState {
  switch (action.type) {
    case actions.CURRENT_CREATING_TRANSACTION_TYPE:
      return {
        ...state,
        currentCreatingTransactionType: action.currentCreatingTransactionType
      };
    case actions.SAVE_TRANSACTION_SUCCESS:
      return {
        ...state,
        lastCreatedTransaction: action.transaction
      };
    case actions.GET_TRANSACTIONS_SUCCESS:
      return {
        ...state,
        transactions: action.transactions
      };
    default:
      return state;
  }
}
