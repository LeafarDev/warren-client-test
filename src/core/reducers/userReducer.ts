import * as actions from '../actionTypes/userActionTypes';

export interface IAccount {
  id: string
  accountNumber: string
  balance: number
  createdAt: string
  updatedAt: string | null
}

export interface IUser {
  id: string
  name: string
  email: string
  accountId: string
  createdAt: string
  updatedAt: string | null
  account: IAccount
}

export interface UserState {
  user: IUser
}

const initialState: UserState = {
  user: {
    id: '',
    name: '',
    email: '',
    accountId: '',
    createdAt: '',
    updatedAt: null,
    account: {
      id: '',
      accountNumber: '',
      balance: 0,
      createdAt: '',
      updatedAt: ''
    }
  }
};

export default function userReducer(
  state: UserState = initialState,
  action: actions.UserAction
): UserState {
  switch (action.type) {
    case actions.SET_USER:
    case actions.GET_USER_SUCCESS:
      return {
        user: action.user
      };
    default:
      return state;
  }
}
