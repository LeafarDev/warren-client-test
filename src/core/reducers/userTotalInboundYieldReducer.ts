import * as actions from '../actionTypes/userTotalInboundYieldActionTypes';

export interface IUserTotalInboundYield {
  total: number
}

export interface UserTotalInboundYieldState {
  userTotalInboundYield: IUserTotalInboundYield;
}

const initialState: UserTotalInboundYieldState = {
  userTotalInboundYield: {
    total: 0
  }
};

export default function userTotalInboundYieldReducer(
  state: UserTotalInboundYieldState = initialState,
  action: actions.UserTotalInboundYieldTypesAction
): UserTotalInboundYieldState {
  switch (action.type) {
    case actions.GET_USER_TOTAL_INBOUND_YIELD_SUCCESS:
      return {
        userTotalInboundYield: action.userTotalInboundYield
      };
    default:
      return state;
  }
}
