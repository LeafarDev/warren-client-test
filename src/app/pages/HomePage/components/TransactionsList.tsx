import React from 'react';
import DataTable from 'react-data-table-component';
import currency from 'currency.js';
import { format, parseISO } from 'date-fns';
import { ITransaction } from '../../../../core/reducers/transactionsReducer';
import { toFixed } from '../../../utils/number';

interface Props {
  transactions: ITransaction[],
}

export const TransactionsList: React.FC<Props> = props => {
  const columns = [
    {
      name: 'Transaction',
      selector: 'type',
      sortable: true
    },
    {
      name: 'Amount',
      selector: (row: any) => currency(toFixed(row.amount, 2), {
        decimal: '.',
        precision: 2
      })
        .format(),
      sortable: true
    }, {
      name: 'Source Account',
      selector: 'sourceAccountNumber'
    }, {
      name: 'Destination Account',
      selector: 'targetAccountNumber'
    },
    {
      name: 'Date',
      selector: (row: any) => format(parseISO(row.createdAt), 'MM-dd-yyyy H:mm:ss'),
      defaultSortAsc: false,
      defaultSortField: true,
      sortable: true,
      right: true
    }
  ];

  const {
    transactions
  } = props;

  return (<DataTable
    title="Transactions"
    columns={columns}
    data={transactions}
  />);
};
