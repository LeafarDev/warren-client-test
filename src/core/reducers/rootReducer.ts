import { combineReducers } from 'redux';
import isLoadingReducer from './isLoadingReducer';
import errorReducer from './errorReducer';
import authReducer from './authReducer';
import transactionsReducer from './transactionsReducer';
import userReducer from './userReducer';
import userTotalInboundYieldReducer from './userTotalInboundYieldReducer';

const rootReducer = combineReducers({
  userTotalInboundYield: userTotalInboundYieldReducer,
  transactions: transactionsReducer,
  auth: authReducer,
  user: userReducer,
  isLoading: isLoadingReducer,
  error: errorReducer
});

export type AppState = ReturnType<typeof rootReducer>;

export default rootReducer;
