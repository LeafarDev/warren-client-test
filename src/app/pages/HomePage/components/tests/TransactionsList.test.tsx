import React from 'react';
import { TransactionsList } from '../TransactionsList';
import { render, screen } from '@testing-library/react';
import { ITransaction } from '../../../../../core/reducers/transactionsReducer';

describe('<TransactionsList />', () => {
  const transactions: ITransaction[] = [
    {
      'id': '848c14cb-79b4-491c-945b-e27c35f8eb30',
      'sourceAccountNumber': '12345678',
      'targetAccountNumber': '65932871',
      'amount': 0.99,
      'type': 'payment',
      'createdAt': '2021-06-02T21:42:02.119Z',
      'updatedAt': null
    },
    {
      'id': '34344fbf-37f3-408e-a290-e65d87f678f2',
      'sourceAccountNumber': '12345678',
      'targetAccountNumber': null,
      'amount': 1000,
      'type': 'withdraw',
      'createdAt': '2021-06-02T21:41:31.196Z',
      'updatedAt': null
    }];

  it('should render and match the snapshot', () => {
    const {
      container: { firstChild }
    } = render(<TransactionsList transactions={transactions}/>);

    expect(firstChild)
      .toMatchSnapshot();
  });

  it('should render the transactions list', () => {

    render(<TransactionsList transactions={transactions}/>);

    expect(screen.getByText('payment'))
      .toBeInTheDocument();
    expect(screen.getByText('withdraw'))
      .toBeInTheDocument();
  });
});
