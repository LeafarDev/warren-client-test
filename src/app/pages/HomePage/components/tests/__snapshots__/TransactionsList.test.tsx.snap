// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TransactionsList /> should render and match the snapshot 1`] = `
<div
  aria-level="1"
  class="sc-fnVZcZ cDmETx rdt_TableHeader"
  role="heading"
>
  <div
    class="sc-fFSPTT fgINsU"
  >
    Transactions
  </div>
  <div
    class="sc-bkbkJK ffLupP"
  />
  <div
    class="sc-hBMUJo fzmYSh"
  >
    <div
      class="sc-bqGGPW diFtgj"
    />
    <div
      class="sc-ksluID efFzQL"
    />
  </div>
</div>
`;
